RecoMuon/SimMuonAlgos/plugins/MuonSimClassifier.cc

```diff
@@ -0,0 +1,574 @@
+
+    reco::MuonToSimCollection recSimColl;
+    reco::SimToMuonCollection simRecColl;
+    edm::LogVerbatim("MuonSimClassifier") <<"\n ***************************************************************** ";
+    edm::LogVerbatim("MuonSimClassifier") <<  " RECO MUON association, type:  "<< trackType_;
+    edm::LogVerbatim("MuonSimClassifier") <<  " ***************************************************************** \n";
+
+    edm::RefToBaseVector<reco::Muon> selMuons;
+    // all muons
+    for (size_t i = 0, n = muons->size(); i < n; ++i) {
+      edm::RefToBase<reco::Muon> rmu = muons->refAt(i);
+      selMuons.push_back(rmu);
+    }
+    
+    edm::RefVector<TrackingParticleCollection> allTPs;
+    for (size_t i = 0, n = trackingParticles->size(); i < n; ++i) {
+        allTPs.push_back(TrackingParticleRef(trackingParticles,i));
+    }
+
+    assoByHits->associateMuons(recSimColl, simRecColl, selMuons, trackType_, allTPs);
+
+    // for global muons without hits on muon detectors, look at the linked standalone muon
+    reco::MuonToSimCollection UpdSTA_recSimColl;
+    reco::SimToMuonCollection UpdSTA_simRecColl;
+    if (trackType_ == reco::GlobalTk) {
+      edm::LogVerbatim("MuonSimClassifier") <<"\n ***************************************************************** ";
+      edm::LogVerbatim("MuonSimClassifier") <<  " STANDALONE (UpdAtVtx) MUON association ";
+      edm::LogVerbatim("MuonSimClassifier") <<  " ***************************************************************** \n";
+      assoByHits->associateMuons(UpdSTA_recSimColl, UpdSTA_simRecColl, selMuons, reco::OuterTk,
+				 allTPs);
+    }
+
+    typedef reco::MuonToSimCollection::const_iterator r2s_it;
+    typedef reco::SimToMuonCollection::const_iterator s2r_it;
+
+    size_t nmu = muons->size();
+    edm::LogVerbatim("MuonSimClassifier") <<"\n There are "<<nmu<<" reco::Muons.";
+
+    std::vector<int> classif(nmu, 0), ext(nmu, 0);
+    std::vector<int> hitsPdgId(nmu, 0), G4processType(nmu, 0), momPdgId(nmu, 0), gmomPdgId(nmu, 0), momStatus(nmu, 0);
+    std::vector<int> flav(nmu, 0), momFlav(nmu, 0), gmomFlav(nmu, 0), hmomFlav(nmu, 0);
+    std::vector<int> tpId(nmu, -1), tpBx(nmu, 999), tpEv(nmu, 999);
+    std::vector<float> signp(nmu, 0.0), pt(nmu, 0.0), eta(nmu, -99.), phi(nmu, -99.);
+    std::vector<float> prodRho(nmu, 0.0), prodZ(nmu, 0.0), momRho(nmu, 0.0), momZ(nmu, 0.0);
+    std::vector<float> tpAssoQuality(nmu, -1);
+
+    std::unique_ptr<reco::GenParticleCollection> secondaries;     // output collection of secondary muons
+    std::map<TrackingParticleRef, int>         tpToSecondaries; // map from tp to (index+1) in output collection
+    std::vector<int> muToPrimary(nmu, -1), muToSecondary(nmu, -1); // map from input into (index) in output, -1 for null
+    if (linkToGenParticles_) secondaries.reset(new reco::GenParticleCollection());
+
+    // loop on reco muons
+    for(size_t i = 0; i < nmu; ++i) {    
+        edm::RefToBase<reco::Muon> mu = muons->refAt(i);
```

Comment:
this can be just a (for) loop over selMuons

Reply:
The loop needs an index to deal with output vectors. Refs are reused now, instead of being recreated.